web/src/main/java/org/springframework/security/web/AroundFilterObservation.java

```diff
@@ -0,0 +1,147 @@
+/*
+ * Copyright 2002-2022 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.springframework.security.web;
+
+import java.util.Arrays;
+import java.util.Iterator;
+import java.util.concurrent.atomic.AtomicReference;
+
+import io.micrometer.observation.Observation;
+import jakarta.servlet.Filter;
+import jakarta.servlet.FilterChain;
+
+interface AroundFilterObservation extends FilterObservation {
+
+	AroundFilterObservation NOOP = new AroundFilterObservation() {
+	};
+
+	static AroundFilterObservation create(Observation before, Observation after) {
+		if (before.isNoop() || after.isNoop()) {
+			return NOOP;
+		}
+		return new SimpleAroundFilterObservation(before, after);
+	}
+
+	default Observation before() {
+		return Observation.NOOP;
+	}
+
+	default Observation after() {
+		return Observation.NOOP;
+	}
+
+	class SimpleAroundFilterObservation implements AroundFilterObservation {
+
+		private final Iterator<Observation> observations;
+
+		private final Observation before;
+
+		private final Observation after;
+
+		private AtomicReference<Observation> currentObservation = new AtomicReference<>(null);
```

Comment:
I'm not sure if it is more convenient or not but `Observation.Scope` has a reference to the `Observation` that you can get via `getCurrentObservation()` so you might get away with one `AtomicReference`.

Reply:
agreed